data_safe_haven/commands/config.py

```diff
@@ -34,9 +35,29 @@ def upload(
 ) -> None:
     """Upload a configuration to the Data Safe Haven context"""
     context = ContextSettings.from_file().assert_context()
+    logger = LoggingSingleton()
+
+    # Create configuration object from file
     with open(file) as config_file:
         config_yaml = config_file.read()
     config = Config.from_yaml(config_yaml)
+
+    # Present diff to user
+    if Config.remote_exists(context):
+        if diff := config.remote_yaml_diff(context):
+            print("".join(diff))
+            if not logger.confirm(
+                (
+                    "Configuration has changed, "
+                    "do you want to overwrite the remote configuration?"
+                ),
+                default_to_yes=False,
+            ):
+                raise typer.Exit()
+        else:
+            print("No changes, won't upload configuration.")
```

Comment:
```suggestion
            logger.info("No changes, won't upload configuration.")
```

Reply:
I prefer logging over printing because:

- you can easily redirect it as needed
- you can enable/disable at different levels of info
- you can turn it off completely if desired

all of which is hard to do with `print` statements.

Reply:
I want to avoid using the logger here because,

- It makes the output very difficult to read
- It bypasses typer stdout so we can't use it in tests (it would break the tests here, and make them a bit pointless as they could only check the return code)

More generally, I think we are sometimes mixing up log information and stdout in commands.
Can we open an issue to sort this?

Reply:
Yes - please open an issue. I'd be interested to know what you mean by bypassing typer stdout - I thought typer didn't do its own I/O but delegated this to rich (which is also what the logger does).

Reply:
> I'd be interested to know what you mean by bypassing typer stdout

I don't entirely understand. However, anything sent to the logger object isn't captured as stdout in tests. I don't know how to inspect any of that.

> In general, I think the only time we should be writing directly to stdout is if we're in a command that is intended to be used as part of a pipe/redirect.

I suppose that is the conversation we need to have, but I think that is incorrect and goes against convention.

I think of stdout as information for the user. Logs are for posterity, for the user/admin.
`pacman -Syu` for example uses stdout to show the user package being downloaded and installed, and hooks being run. You wouldn't pipe that to anything else. Logs of transactions are also written to logs in `/var/logs/pacman.log`.

Or, a more common/mundane example, you would rarely pipe `<command> --help` to another command, but it would be very odd to not send that to stdout.

Reply:
#1891